hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/publishing/end/MessageErrorProcessor.java

```diff
@@ -0,0 +1,129 @@
+package pl.allegro.tech.hermes.frontend.publishing.end;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.util.Headers;
+import io.undertow.util.HttpString;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import pl.allegro.tech.hermes.api.ErrorCode;
+import pl.allegro.tech.hermes.api.ErrorDescription;
+import pl.allegro.tech.hermes.api.Topic;
+import pl.allegro.tech.hermes.frontend.publishing.message.MessageIdGenerator;
+import pl.allegro.tech.hermes.tracker.frontend.Trackers;
+
+import javax.inject.Inject;
+import javax.ws.rs.core.MediaType;
+import java.io.IOException;
+
+import static org.apache.commons.lang3.exception.ExceptionUtils.getRootCauseMessage;
+import static pl.allegro.tech.hermes.api.ErrorCode.GROUP_NOT_EXISTS;
+import static pl.allegro.tech.hermes.api.ErrorCode.INTERNAL_ERROR;
+import static pl.allegro.tech.hermes.api.ErrorCode.TIMEOUT;
+import static pl.allegro.tech.hermes.api.ErrorCode.TOPIC_NOT_EXISTS;
+import static pl.allegro.tech.hermes.api.ErrorCode.VALIDATION_ERROR;
+import static pl.allegro.tech.hermes.api.ErrorDescription.error;
+import static pl.allegro.tech.hermes.common.http.MessageMetadataHeaders.MESSAGE_ID;
+
+public class MessageErrorProcessor {
+    private static final Logger logger = LoggerFactory.getLogger(MessageErrorProcessor.class);
+    private final ObjectMapper objectMapper;
+    private final Trackers trackers;
+    private final HttpString messageIdHeader = new HttpString(MESSAGE_ID.getName());
+
+    @Inject
+    public MessageErrorProcessor(ObjectMapper objectMapper, Trackers trackers) {
+        this.objectMapper = objectMapper;
+        this.trackers = trackers;
+    }
+
+    public void missingTopicGroup(HttpServerExchange exchange, String qualifiedTopicName) {
+        sendErrorResponseQuietly(
+                exchange,
+                error("Missing valid topic group in path. Found " + qualifiedTopicName, GROUP_NOT_EXISTS),
+                MessageIdGenerator.generate());
+    }
+
+    public void nonExistentTopic(HttpServerExchange exchange, String qualifiedTopicName) {
+        sendErrorResponseQuietly(
+                exchange,
+                error("Topic not found: " + qualifiedTopicName, TOPIC_NOT_EXISTS),
+                MessageIdGenerator.generate());
+    }
+
+    public void readException(HttpServerExchange exchange, Topic topic, String messageId, IOException exception) {
+        log(error("Error while reading message. " + getRootCauseMessage(exception), INTERNAL_ERROR), topic, messageId, exchange.getHostAndPort());
+        // client closed connection, there is no way to send him a response
```

Comment:
fixed, from now on frontend checks whether connection is open, if it is then response is returned.